File: oneflux_steps/ustar_cp_refactor_wip/fcr2Calc.m
	function r2=fcr2Calc(y,yHat);    
   
   % This function calculates the r^2 measure (coeffecient of determination)
   % for a pair of data `y` and some predicted `yHat`.
  
   n=length(y); ym=mean(y); 
   
   % explained sum of squares
   SSreg=sum((yHat-ym).^2); 

   % total sum of squares
   SStotal=sum((y-ym).^2); 

   rmse=sum((yHat-y).^2);  % Note this is redundant

   % coeffecient of determination
   r2=SSreg/SStotal; 
   
% % %    sprintf('%7.4f %7.4f %7.4f ',[SSreg SStotal r2 rmse]) 
   

File: oneflux_steps/ustar_cp_python/fcr2Calc.py
import numpy as np
from oneflux_steps.ustar_cp_python.utils import squeeze

def fcr2Calc(y=None, yHat=None):
    """
    Calculate the coefficient of determination (R^2) for the given observed and predicted values.

    Parameters:
    y (array-like): Observed values.
    yHat (array-like): Predicted values.

    Returns:
    float: The coefficient of determination (R^2), which indicates the proportion of the variance in the dependent variable that is predictable from the independent variable(s).

    Notes:
    - SSreg (Sum of Squares of the regression) is calculated as the sum of the squared differences between the predicted values and the mean of the observed values.
    - SStotal (Total Sum of Squares) is calculated as the sum of the squared differences between the observed values and the mean of the observed values.
    - RMSE (Root Mean Square Error) is calculated as the sum of the squared differences between the predicted values and the observed values.
    - R^2 is then calculated as the ratio of SSreg to SStotal.

    Example:
    >>> y = [3, -0.5, 2, 7]
    >>> yHat = [2.5, 0.0, 2, 8]
    >>> fcr2Calc(y, yHat)
    1.214132762312634
    """
    n = len(y)

    # Calculate the mean of the observed values.
    ym = np.mean(squeeze(np.asarray(y)), axis=0)
    # SSreg ((explained) Sum of Squares of the regression) 
    SSreg = np.sum((yHat - ym) ** 2)
    # SStotal (Total Sum of Squares)
    SStotal = np.sum((y - ym) ** 2)

    # RMSE (Root Mean Square Error)
    # TODO: can be remove as it is uunused
    rmse = np.sum((yHat - y) ** 2)

    # Calculate the coefficient of determination
    r2 = SSreg / SStotal
    return r2
